### ISSUE TITLE
Django wtih slim handler causes random timeouts

### ISSUE THREAD

**Issue by svechinsky** [Miserlou/Zappa#1495]
## Context
I'm using python3.6 and django-rest-framework to create an api. ever since I turned slim handler on I noticed that API gateway occasionally times out (after 29 seconds as per the hard limit) and the lambda function (managed by zappa) only returns after 60+ seconds.

## Expected Behavior
Uniform behavior across executions.
## Actual Behavior
Endpoint occasionally times out.

## Your Environment
`pip freeze` output:
`aiodns==1.1.1
aiohttp==3.0.1
alabaster==0.7.10
argcomplete==1.9.2
asn1crypto==0.24.0
astroid==1.5.3
async-timeout==2.0.1
attrs==17.4.0
autobahn==18.4.1
Automat==0.6.0
autopep8==1.3.4
Babel==2.5.1
base58==0.2.4
boto3==1.7.10
botocore==1.10.10
cchardet==2.1.1
ccxt==1.11.171
certifi==2018.4.16
cffi==1.11.4
cfn-flip==1.0.3
chardet==3.0.4
click==6.7
constantly==15.1.0
cryptography==2.1.4
dateparser==0.7.0
decorator==4.1.2
deux==1.2.0
Django==1.11.7
django-cors-headers==2.1.0
django-oauth-toolkit==1.0.0
django-otp==0.4.2
django-storages==1.6.5
djangorestframework==3.7.3
djangorestframework-jwt==1.11.0
docopt==0.6.2
docutils==0.14
drf-jwt-2fa==0.3.0
drf-writable-nested==0.4.1
durationpy==0.5
future==0.16.0
hjson==3.0.1
hyperlink==18.0.0
idna==2.6
idna-ssl==1.0.1
imagesize==0.7.1
incremental==17.5.0
ipython==6.2.1
ipython-genutils==0.2.0
isort==4.2.15
jedi==0.11.0
Jinja2==2.10
jmespath==0.9.3
kappa==0.6.0
lambda-packages==0.19.0
lazy-object-proxy==1.3.1
MarkupSafe==1.0
mccabe==0.6.1
multidict==4.1.0
mypy==0.550
oauthlib==2.0.6
parso==0.1.0
pep8==1.7.1
pexpect==4.3.0
pickleshare==0.7.4
placebo==0.8.1
prompt-toolkit==1.0.15
psutil==5.4.1
psycopg2==2.7.3.2
ptipython==1.0.0
ptpython==0.41
ptyprocess==0.5.2
pyasn1==0.4.2
pyasn1-modules==0.2.1
pycares==2.3.0
pycodestyle==2.3.1
pycparser==2.18
Pygments==2.2.0
PyJWT==1.5.3
pylint==1.7.4
pymongo==3.5.1
pyOpenSSL==17.5.0
PySocks==1.6.8
python-binance==0.6.8
python-dateutil==2.6.1
python-dotenv==0.7.1
python-http-client==3.0.0
python-slugify==1.2.4
pytimeparse==1.1.7
pytz==2017.3
PyYAML==3.12
regex==2018.2.21
requests==2.18.4
s3transfer==0.1.13
sendgrid==5.3.0
service-identity==17.0.0
simplegeneric==0.8.1
six==1.11.0
snowballstemmer==1.2.1
Sphinx==1.6.5
sphinxcontrib-websupport==1.0.1
toml==0.9.4
tqdm==4.19.1
traitlets==4.3.2
troposphere==2.2.1
twilio==6.10.2
Twisted==17.9.0
txaio==2.10.0
typed-ast==1.1.0
tzlocal==1.5.1
Unidecode==1.0.22
urllib3==1.22
wcwidth==0.1.7
websocket-client==0.44.0
Werkzeug==0.12
wrapt==1.10.11
wsgi-request-logger==0.4.6
yarl==1.1.0
zappa==0.45.1
zope.interface==4.5.0
`

`zappa_settings.py`: 
`{
    "staging": {
        "aws_region": "us-east-1",
        "django_settings": "xxxxxxx.settings.staging",
        "profile_name": "xxxxxx",
        "project_name": "xxxxx",
        "runtime": "python3.6",
        "s3_bucket": "zappa-xxxxx",
        "vpc_config": { 
            "SubnetIds": [ "subnet-xxxxx", "subnet-xxxx" ], 
            "SecurityGroupIds": [ "sg-xxxxx" ]
        },
        "slim_handler":true,
        "cors":true,
        "timeout_seconds": 150,
        "include":[]
    }
`

**Comment by paulclarkaranz** [comment]
> occasionally times out

Maybe you can check the logs and see if these timing out requests are ones that always have the re-instancing message before them. That is the only part of the process I could think of slim handler effecting

**Comment by svechinsky** [comment]
Brilliant, spot on.
Any way to easily fix this?

**Comment by svechinsky** [comment]
Additional info:
I've fiddled around with the logs before but started looking at the timstamps after you mentioned it.
I think the problem is here:

> [1525687012196] [DEBUG] 2018-05-07T09:56:52.195Z f182b4e8-51dc-11e8-b70b-4d15a433e5ef Response: {'ResponseMetadata': {'RequestId': '1BE94793E17AFDE4', 'HostId': 'vpUjUEekQ9kX6lRG+P0oQbc92ZqSqWUyHUSREqIHdoxZYkUBy/UAGpx/iQr3f91flv8LkdEG6t8=', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amz-id-2': 'vpUjUEekQ9kX6lRG+P0oQbc92ZqSqWUyHUSREqIHdoxZYkUBy/UAGpx/iQr3f91flv8LkdEG6t8=', 'x-amz-request-id': '1BE94793E17AFDE4', 'date': 'Mon, 07 May 2018 09:56:53 GMT', 'last-modified': 'Mon, 07 May 2018 08:33:10 GMT', 'etag': '"52dbd56a5bb635a84d15b64101d3d7ca-8"', 'accept-ranges': 'bytes', 'content-type': 'binary/octet-stream', 'content-length': '58736875', 'server': 'AmazonS3'}, 'RetryAttempts': 0}, 'AcceptRanges': 'bytes', 'LastModified': datetime.datetime(2018, 5, 7, 8, 33, 10, tzinfo=tzutc()), 'ContentLength': 58736875, 'ETag': '"52dbd56a5bb635a84d15b64101d3d7ca-8"', 'ContentType': 'binary/octet-stream', 'Metadata': {}, 'Body': <botocore.response.StreamingBody object at 0x7f4a90985a20>}
> [1525687036266] Failed to find library libmysqlclient.so.18 ...right filename?
> [1525687037978] Failed to find library libmysqlclient.so.18 ...right filename?
> [1525687038158] Failed to find library libmysqlclient.so.18 ...right filename?
> 

Timestamp difference here is the biggest.
Maybe this is caused by libmysql?
https://github.com/Miserlou/Zappa/pull/789

The keep warm handler only warms one instance which fails when trying to do multiple requests.
Is there any way to get past this?

**Comment by svechinsky** [comment]
I did find an easy fix after all.
Simply don't use slim handler. 
I accidentally stumbled across [this](https://hackernoon.com/exploring-the-aws-lambda-deployment-limits-9a8384b0bec3) post and apparently if you upload your code to lambda through S3 the limit is actually 250MB of unzipped code.

**Comment by KophyGiddie** [comment]
It has nothing to do with the slim handler or size of your code(of course there is a limit of 512mb of the uncompressed code as stated in the AWS docs) but rather if your code is running in the VPC. Your subnet configuration can get a little messy and thats what causes the issue. find the correct setup here https://aws.amazon.com/premiumsupport/knowledge-center/internet-access-lambda-function/. Also one more thing. In your route table make sure you enable an s3 endpoint that matches your region and that should solve it permanently.

**Comment by svechinsky** [comment]
So I had to go back to using slim handler due to an increase in our codebase and the problem seems to still exist.

It might have something to do with the S3 though I doubt it since the endpoint is available and there are routes in every subnet associated with it.

Regarding the NAT gateway setup, the only issue that might be causing this is that the NAT gateway we have is positioned on one availability zone where the lambda has subnets on 3 zones.

But if that's the issue then why did the problem stop when I disabled slim handler?

**Comment by svechinsky** [comment]
For anyone that's interested I found another solution:
The main time delay (around 15s) was in fetching the code from S3 on initialization.
An additional delay was loading the libraries (around 10s).
This left around 5-6 sec for the code to execute.

From here: https://github.com/Miserlou/Zappa/issues/1974
I learned that lambda provides some kind of resource boost when initializing the container so I called the initialization function (the one that fetches the code) as part of the file loading instead of as part of the handler only. 
This solved the issue.
Did so by adding a LambdaHandler() call at the bottom of the file
